ComputerCraft Integration

caelwarner opened the pull request:
## Adding integration for CC: Tweaked

Current tile entities added as peripherals:
- **Display Link:** Write to single line and multi line display targets
- **Rotation Speed Controller:** Set and get target rotation speed
- **Speedometer:** Get current rotation speed
- **Stressometer:** Get current stress level and network stress capacity
- **Sequenced Gearshift:** Rotate by an angle, or move a piston/pulley/gantry by a distance
- **Train Station:** Set and get station name, train name and train schedule. Assemble and disassemble trains

Documentation for the added peripherals is [here](https://github.com/Creators-of-Create/Create/tree/mc1.18/computercraft/wiki)

## Implementation
- Tile entities each have their own Peripheral class. The Peripheral class defines the API for Lua and sets the name of the peripheral. Each peripheral has access to a reference of its tile entity
- The Peripheral class extends `SyncedPeripheral` which implements `IPeripheral`. `SyncedPeripheral`'s keep track of how many computers are connected to them at one time.
- Peripherals that have a GUI can use `ComputerScreen` to show to the user that they are currently be controlled by a computer.
- Tile entities that have peripherals can use `ComputerBehaviour` to add the required peripheral capability.

caelwarner committed:
ComputerCraft integration for Display Links

- CC computers can now control display links through a variety of functions
- Added ComputerControllable interface to define a tile entity as controllable by CC computers
- Added CC: Tweaked soft dependency

caelwarner committed:
Refactored peripheralHandler to peripheral

- peripheralHandler was the wrong name. It's just a peripheral.
- Changed peripheral type from "cdl" to "Create_DisplayLink"
- Added equals function to DisplayLinkPeripheral

caelwarner committed:
Added Speedometer and Stressometer as peripherals

- Speedometer can get current speed
- Stressometer can get current stress level as well as network stress capacity
- Made GaugeTileEntity abstract

caelwarner committed:
Added Rotation Speed Controller and Sequenced Gearshift as peripherals

- Rotation Speed Controller can get and set targetSpeed
- Sequenced Gearshift can rotate by a certain angle and move a certain distance

caelwarner committed:
Added invalidateCaps

- Changed setRemoved to invalidateCaps. I don't know why I wasn't just using invalidateCaps from the beginning

caelwarner committed:
Added PeripheralBase

caelwarner committed:
Sequenced Gearshift screen "greys out" when being controlled by a computer

- This is to stop players from trying to using both the builtin sequencing and a computer to control the Sequenced Gearshift at the same time, leading to undefined behaviour
- The "greyed out" screen should have a message added explaining why it's greyed out.
- Added ComputerControllable#isComputerControlled to check if a tile entity is connected to a modem

SquidDev commented:
It may also be worth chatting with the developer of [CC:C Bridge](https://github.com/tweaked-programs/cccbridge), and seeing if there's anything there they think is worth adding to this PR.

caelwarner committed:
Changed method of checking if a computer attached

- After talking with SquidDev from CC: Tweaked I've changed to monitoring IPeripheral#attach and IPeripheral#detach for changes in the number of computers connected to the network, then updating the client using AttachedComputerPacket
- This works with wired full modems, wired cabled modems and directly connected computers
- Added SyncedPeripheralBase and SyncedComputerControllable for TE's and peripherals that want to be aware of attached computers

caelwarner committed:
Changed Display Link Peripheral API

- Changed the Display Link Peripheral API to be more in line with the Monitor API
- Added write, setCursorPos, getCursorPos, getSize, isColor, isColour, clearLine
- Removed void writeLine, setLine, writeLines, setLines

Max-CGN commented:
a way for a computer at a train station to assemble and disassemble trains would be nice

caelwarner committed:
Added Train Station as peripheral

- Train station can set a new auto-schedule for the train currently at the station
- Added CreateLuaTable to add helper functions for working with lua tables
- Added StringHelper util class to convert snake case to camel case

caelwarner committed:
Moved all peripheral classes to computercraft.peripherals package

caelwarner committed:
Created ComputerBehaviour behaviour

- ComputerBehaviour replaces ComputerControllable and SyncedComputerControllable

caelwarner committed:
Made LuaFunction's final

caelwarner committed:
Refactored PeripheralBase to SyncedPeripheral

caelwarner committed:
Added ComputerScreen

- ComputerScreen shows that tile entity currently has computers attached and therefore cannot be controlled manually

caelwarner committed:
Serialize hasAttachedComputer in ComputerBehaviour

- This eliminates some edge cases were peripherals don't realize they're being controlled by a computer on a world save and load

kristibektashi commented:
Any progress in this?

caelwarner committed:
Extended train station peripheral API

- Train station peripherals can now assemble and disassemble trains, check if the station is in assembly mode, set the assembly mode of the station, get and change the station name, check if a train is present at the station and get and change the currently present train name.
- Refactored StationEditPacket. Moved most of the logic that was previously in StationEditPacket to StationTileEntity. This allows us to call this logic without having to send a packet.
- Made Train#owner nullable. This is needed so that computers can assemble trains. All Train#owner is currently used for is to display the train status to the correct play.

caelwarner committed:
Added getSchedule to train station lua API

- Added getSchedule which serializes the currently present train's schedule into a lua table
- Refactored StationPeripheral#setSchedule to use a more generic method of serializing NBT tags to lua tables
- Moved schedule entry special data from root tag to "Data"
- Added StringHelper#camelCaseToSnakeCase
- Added variety of put methods to CreateLuaTable

caelwarner committed:
Added isTrainImminent and hasSchedule to train station API

- Added isTrainImminent to check if a train is incoming to the station and hasSchedule to check if the currently present train has a schedule
- Added StationPeripheral#getTrainOrThrow to consolidate repetitive null checks

caelwarner committed:
Updated DisplayLinkPeripheral#write to move cursor to the end of the text

- This change was made to be more inline with ComputerCraft's builtin display API

caelwarner committed:
Added documentation for most peripherals

- Lua documentation has been added for all peripherals except the train station (not looking forward to writing that one)
- This documentation will be added to the GitHub wiki pages

caelwarner commented:
> Any progress in this?

Yes @kristibektashi, I'm approaching the end of this integration. You can view the documentation for the added peripherals [here](https://github.com/Creators-of-Create/Create/tree/mc1.18/computercraft/wiki). Except for the train station which doesn't have documentation yet, though you can view the source code for that [here](https://github.com/Creators-of-Create/Create/blob/mc1.18/computercraft/src/main/java/com/simibubi/create/compat/computercraft/peripherals/StationPeripheral.java)

caelwarner committed:
Added isTrainEnroute to station peripheral API

- isTrainEnroute checks if a train is currently navigating to the station
- Reworded null station exception to "station is not connected to a track"
- Refactored StationPeripheral#inAssemblyMode to StationPeripheral#isInAssemblyMode
- Added a check to StationPeripheral#disassemble to make sure the station isn't in assembly mode

caelwarner committed:
Added documentation for train station peripherals and train schedules

- Added in depth documentation for working with train stations and train schedules in Lua
- Fixed small formatting issues in Lua-Rotation-Speed-Controller.md and Lua-Sequenced-Gearshift.md

caelwarner commented:
Documentation has been added for all peripherals now!

caelwarner committed:
Added computer to display source ponder tag

- Added advanced computer to display source ponder tag
- Added missing lang entry for computer display source

ChristianLW committed:
Small tweaks to the wiki pages

zyxkad commented:
At <https://github.com/Creators-of-Create/Create/blob/mc1.18/computercraft/wiki/Lua-Sequenced-Gearshift.md#rotateangle-modifier>, I would like to throw a lua error when the speed is out of the range [-2, 2], it will keep the users from doing unexpected behavior.

And it doesn't explain what will happen if we call the methods when the seq gearshift is in progress (or running it's own program), I suggest we just add another block for computer control? or add a process for the gearshift that clearly described this block is used for computer

simibubi committed:
Merge pull request #4650 from ChristianLW/mc1.18/computercraft

Small tweaks to the wiki pages for ComputerCraft integration

simibubi committed:
Compilation dependency toggle

simibubi commented:
Merged in https://github.com/Creators-of-Create/Create/commit/3ad4195dd6865fb0994c37f9d95555aed35373e9

Ch1216 commented:
It be neat if for the next update (if there is gonna be a next one) it was compatible with elevator pulleys, especially since compared to other blocks, you cant modify them with just rotation (in which case you could just control a rotation speed controller or sequence gearshift to get the job done)
the methods could be: getFloor() [Outputs the closest floor the elevator is in], and setFloor(number) [calls the elevator to a specific floor]
